Geometry/HGCalGeometry/src/HGCalGeometry.cc

```diff
@@ -260,12 +253,24 @@ unsigned int HGCalGeometry::sizeForDenseIndex() const {
   return topology().totalGeomModules();
 }
 
-const CaloCellGeometry* HGCalGeometry::cellGeomPtr(uint32_t index) const {
+std::shared_ptr<const CaloCellGeometry> HGCalGeometry::cellGeomPtr(uint32_t index) const {
+  if ((index >= m_cellVec.size()) || (m_validGeomIds[index].rawId() == 0)) 
+    return nullptr;
+  static const auto do_not_delete = [](const void*){};
+  auto cell = std::shared_ptr<const CaloCellGeometry>(&m_cellVec[index],do_not_delete);
+  if (nullptr == cell->param()) return nullptr;
+  return cell;
+}
+
+std::shared_ptr<const CaloCellGeometry> HGCalGeometry::cellGeomPtr(uint32_t index, const GlobalPoint& pos) const {
   if ((index >= m_cellVec.size()) || (m_validGeomIds[index].rawId() == 0)) 
     return nullptr;
-  const CaloCellGeometry* cell ( &m_cellVec[ index ] ) ;
+  if (pos == GlobalPoint()) return cellGeomPtr(index);
+  FlatTrd* newcell = new FlatTrd(m_cellVec[index]);
```

Comment:
Like above, better to change this to
`auto cell = std::make_shared<FlatTrd>(m_cellVec[index]);`